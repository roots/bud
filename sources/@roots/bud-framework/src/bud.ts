import {bind} from '@roots/bud-support/decorators'
import isFunction from '@roots/bud-support/lodash/isFunction'
import isNull from '@roots/bud-support/lodash/isNull'
import isString from '@roots/bud-support/lodash/isString'
import isUndefined from '@roots/bud-support/lodash/isUndefined'

import {bootstrap, LIFECYCLE_EVENT_MAP} from './lifecycle/bootstrap.js'
import type * as methods from './methods/index.js'
import type {Module} from './module.js'
import type {Service} from './service.js'
import type ConsoleBuffer from './services/console.js'
import type FS from './services/fs.js'
import type * as Options from './types/options/index.js'
import type * as Registry from './types/registry/index.js'
import type Hooks from './types/services/hooks/index.js'
import type * as Services from './types/services/index.js'
import Value from './value.js'

/**
 * Framework abstract
 */
export class Bud {
  /**
   * Context
   *
   * @public
   */
  public context:
    | Options.CommandContext
    | Options.CLIContext
    | Options.Context

  public isCLI(): this is Bud & {context: Options.CommandContext} {
    return `args` in this.context
  }

  /**
   * Implementation
   *
   * @public
   */
  public implementation: new () => Bud

  /**
   * Compilation mode
   *
   * @remarks
   * Either `production` or `development`.
   *
   * @readonly
   * @defaultValue `production`
   * @public
   */
  public get mode(): `development` | `production` {
    return this.context.mode ?? `production`
  }

  /**
   * Name
   *
   * @readonly
   * @public
   */
  public get label() {
    return this.context.label
  }

  /**
   * Parent {@link Bud} instance
   *
   * @readonly
   * @public
   */
  public get root() {
    return this.context.root ?? this
  }

  /**
   * True when {@link Bud.mode} is `production`
   *
   * @public
   */
  public get isProduction(): boolean {
    return this.mode == `production`
  }

  /**
   * True when {@link Bud.mode} is `development`
   *
   * @public
   */
  public get isDevelopment(): boolean {
    return this.mode == `development`
  }

  /**
   * True when current instance is the parent instance
   *
   * @readonly
   * @public
   */
  public get isRoot(): boolean {
    return this.root.label === this.label
  }

  /**
   * True when current instance is a child instance
   *
   * @readonly
   * @public
   */
  public get isChild(): boolean {
    return this.root.label !== this.label
  }

  /**
   * {@link Bud} instances
   *
   * @public
   */
  public children?: Record<string, Bud>

  /**
   * True when child compilers
   *
   * @readonly
   * @public
   */
  public get hasChildren(): boolean {
    return (
      !isUndefined(this.children) &&
      !isNull(this.children) &&
      Object.entries(this.children).length > 0
    )
  }

  public consoleBuffer: ConsoleBuffer

  public fs: FS

  public module: Module

  public services: Array<string> = []

  public api: Services.Api

  public build: Services.Build.Service

  public cache: Services.Cache.Service

  public compiler: Services.Compiler.Service

  public dashboard: Services.Dashboard.Service

  public env: Services.Env

  public extensions: Services.Extensions.Service

  public hooks: Hooks

  public project: Services.Project.Service

  public server: Services.Server.Service

  public declare after: methods.after

  public declare maybeCall: methods.maybeCall

  public declare close: methods.close

  public declare container: methods.container

  public declare get: methods.get

  public declare glob: methods.glob

  public declare globSync: methods.globSync

  public declare path: methods.path

  public declare pipe: methods.pipe

  public declare processConfigs: methods.processConfigs

  public declare publicPath: methods.publicPath

  public declare relPath: methods.relPath

  public declare run: methods.run

  public declare setPath: methods.setPath

  public declare setPublicPath: methods.setPublicPath

  public declare sequence: methods.sequence

  public declare sequenceSync: methods.sequenceSync

  public declare sh: methods.sh

  public declare tap: methods.tap

  public declare tapAsync: methods.tapAsync

  public declare when: methods.when

  public declare bindMethod: methods.bindMethod

  /**
   * @deprecated - use {@link bud.fs.json | bud.fs.json}
   */
  public json: FS['json']

  /**
   * @deprecated - use {@link bud.fs.yml | bud.fs.yml}
   */
  public yml: FS['yml']

  /**
   * Value helper
   *
   * @public
   */
  public value = Value

  /**
   * Creates a child with `bud.create` but returns the parent instance
   *
   * @public
   * @decorator `@bind`
   */
  @bind
  public async make(
    request: Partial<Options.Context> | string,
    tap?: (app: Bud) => Promise<unknown>,
  ) {
    if (!this.isRoot)
      return this.fatal(
        new Error(
          `Child instances should be produced from the root context`,
        ),
      )

    const context: Options.Context = isString(request)
      ? {...this.context, label: request, root: this}
      : {...this.context, ...request, root: this}

    if (
      this.isCLI() &&
      !isUndefined(this.context.args.filter) &&
      !this.context.args.filter.includes(context.label)
    ) {
      this.log(
        `skipping child instance based on --filter flag:`,
        context.label,
      )
      return this
    }

    if (this.children && this.children[context.label]) {
      this.log(`returning requested child instance:`, context.label)
      return this.get(context.label)
    }

    this.log(`instantiating new bud instance`)
    const child = await new Bud().lifecycle(context)

    if (!this.children) this.children = {[context.label]: child}
    else this.children[context.label] = child

    this.get(context.label).hooks.on(
      `build.dependencies`,
      typeof request !== `string` && request.dependsOn
        ? request.dependsOn
        : Object.values(this.children)
            .map(({label}) => label)
            .filter(label => label !== context.label),
    )

    if (tap) await tap(this.get(context.label))
    await this.get(context.label)?.api.processQueue()

    return this
  }

  @bind
  public async lifecycle(context: Options.Context): Promise<Bud> {
    await bootstrap.bind(this)({...context})

    Object.entries(LIFECYCLE_EVENT_MAP).map(
      ([eventHandle, callbackName]: [
        keyof Registry.EventsStore,
        keyof Service,
      ]) =>
        this.services
          .map(service => [service, this[service]])
          .map(([label, service]) => {
            if (!service) {
              this.error(`service not found: ${label}`, this.services)
            }

            if (!isFunction(service[callbackName])) return

            this.hooks.action(
              eventHandle,
              service[callbackName].bind(service),
            )
            this.success(
              `registered service callback:`,
              `${label}.${callbackName}`,
            )
          }),
    )

    await [
      `init`,
      `bootstrap`,
      `bootstrapped`,
      `register`,
      `registered`,
      `boot`,
      `booted`,
    ].reduce(async (promised, event: keyof Registry.EventsStore) => {
      await promised
      await this.hooks.fire(event)
    }, Promise.resolve())

    return this
  }

  /**
   * Log a message
   *
   * @public
   * @decorator `@bind`
   */
  @bind
  public log(...messages: any[]) {
    this.context.logger.log(...messages)

    return this
  }

  /**
   * Log an `info` level message
   *
   * @public
   * @decorator `@bind`
   */
  @bind
  public info(...messages: any[]) {
    this.context.logger.info(...messages)

    return this
  }

  /**
   * Log a `success` level message
   *
   * @public
   * @decorator `@bind`
   */
  @bind
  public success(...messages: any[]) {
    this.context.logger.success(...messages)
    return this
  }

  /**
   * Log a `warning` level message
   *
   * @public
   * @decorator `@bind`
   */
  @bind
  public warn(...messages: any[]) {
    this.context.logger.warn(...messages)

    return this
  }

  /**
   * Log an error.
   *
   * @remarks
   * In `production` this error is treated as fatal
   * and will kill the process.
   *
   * @public
   * @decorator `@bind`
   */
  @bind
  public error(...messages: Array<any>): Bud {
    this.context.logger.error(...messages)
    return this
  }

  /**
   * Log and display an error.
   *
   * @remarks
   * This will always kill the process
   *
   * @public
   * @decorator `@bind`
   * @throws fatal error
   */
  @bind
  public fatal(error: Error) {
    this.context.logger.fatal(error)
    return this
  }
}
