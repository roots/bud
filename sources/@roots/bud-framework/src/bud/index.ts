import type {
  Api,
  Build,
  Cache,
  Compiler,
  Context,
  Dashboard,
  Env,
  Extensions,
  Hooks,
  Project,
  Server,
  Service,
} from '@roots/bud-framework'

import methods from '@roots/bud-framework/methods'
import {bind} from '@roots/bud-support/decorators/bind'
import {BudError, InputError} from '@roots/bud-support/errors'
import isFunction from '@roots/bud-support/isFunction'
import isString from '@roots/bud-support/isString'
import isUndefined from '@roots/bud-support/isUndefined'
import logger from '@roots/bud-support/logger'

import type {FS} from '../fs.js'
import type {Module} from '../module.js'
import type {Notifier} from '../notifier.js'
import type {EventsStore} from '../registry/index.js'

import {bootstrap} from '../bootstrap.js'

/**
 * Bud core class
 */
export class Bud {
  /**
   * Promised tasks
   */
  public promised: Array<(bud: Bud) => Promise<any>> = []

  public declare addConfig: typeof methods.addConfig

  public declare after: typeof methods.after

  public declare api: Service & Api

  public declare bindFacade: typeof methods.bindFacade

  public declare build: Service & Build

  public declare cache: Service & Cache

  /**
   * {@link Bud} instances
   */
  public declare children: Record<string, Bud>

  public declare close: typeof methods.close

  public declare compiler?: Service & Compiler

  public declare container: typeof methods.container

  public declare context: Context

  public declare dashboard: Service & Dashboard

  public declare env: Service & Env

  public declare extensions: Service & Extensions

  public declare fs: FS

  public declare get: typeof methods.get

  public declare glob: typeof methods.glob

  public declare globSync: typeof methods.globSync

  public declare hooks: Service & Hooks

  /**
   * {@link Bud} Implementation
   * @internal
   */
  public declare implementation: new () => Bud

  /**
   * @deprecated Use {@link bud.fs}
   * @readonly
   */
  public declare json: FS['json']

  public declare maybeCall: typeof methods.maybeCall

  public declare module: Module

  public declare notifier: Notifier

  public declare path: typeof methods.path

  public declare pipe: typeof methods.pipe

  public declare processConfigs: typeof methods.processConfigs

  public declare project: Project

  public declare publicPath: typeof methods.publicPath

  public declare relPath: typeof methods.relPath

  public declare run: typeof methods.run

  public declare sequence: typeof methods.sequence

  public declare sequenceSync: typeof methods.sequenceSync

  public declare server?: Service & Server

  public declare setPath: typeof methods.setPath

  public declare setPublicPath: typeof methods.setPublicPath

  public declare sh: typeof methods.sh

  public declare tap: typeof methods.tap

  public declare tapAsync: typeof methods.tapAsync

  public declare when: typeof methods.when

  /**
   * @deprecated Use {@link Bud.fs}
   * @readonly
   */
  public declare yml: FS['yml']

  /**
   * True when child compilers
   * @readonly
   */
  public get hasChildren(): boolean {
    return this.children && Object.entries(this.children).length > 0
  }

  /**
   * True when current instance is a child instance
   * @readonly
   */
  public get isChild(): boolean {
    return this.root?.context?.label !== this.context?.label
  }

  /**
   * True when {@link Bud.mode} is `development`
   * @readonly
   */
  public get isDevelopment(): boolean {
    return this.mode === `development`
  }

  /**
   * True when {@link Bud.mode} is `production`
   * @readonly
   */
  public get isProduction(): boolean {
    return this.mode === `production`
  }

  /**
   * True when current instance is the parent instance
   * @readonly
   */
  public get isRoot(): boolean {
    return this.root?.context?.label === this.context?.label
  }

  /**
   * Label
   * @readonly
   */
  public get label() {
    return this.context?.label ?? `bud`
  }

  /**
   * Compilation mode
   *
   * @remarks
   * Either `production` or `development`.
   *
   * @readonly
   * @defaultValue `production`
   */
  public get mode(): `development` | `production` {
    return this.context?.mode ?? `production`
  }

  /**
   * Constructor
   */
  public constructor(context?: Context) {
    if (context) this.context = {...context}

    this.set(`implementation`, this.constructor as any)

    Object.entries(methods).map(([k, v]) =>
      this.set(k as any, v.bind(this)),
    )
  }

  /**
   * Boot application services
   */
  @bind
  public async boot() {
    await this.executeServiceCallbacks(`boot`).catch(this.catch)
  }

  /**
   * Bootstrap the application
   */
  @bind
  public async bootstrap() {
    await this.executeServiceCallbacks(`bootstrap`).catch(this.catch)
  }

  /**
   * Error handler
   */
  @bind
  public catch(error: Error): never {
    const normalizedError = BudError.normalize(error)
    if (!normalizedError.instance && this?.isChild)
      normalizedError.instance = this.label
    throw normalizedError
  }

  /**
   * Execute service callbacks for a given stage
   * @internal
   */
  @bind
  public async executeServiceCallbacks(
    stage: `${keyof EventsStore & string}`,
  ): Promise<void> {
    await this.resolvePromises().catch(this.catch)
    await this.hooks.fire(stage, this)
  }

  /**
   * Bud initialize
   */
  @bind
  public async initialize(context?: Context): Promise<Bud> {
    if (context) this.context = {...(this.context ?? {}), ...context}

    await bootstrap(this).catch(this.catch)

    await this.bootstrap().catch(this.catch)
    await this.register().catch(this.catch)
    await this.boot().catch(this.catch)
    await this.executeServiceCallbacks(`config.before`).catch(this.catch)

    return this
  }

  /**
   * Creates a child and returns the parent instance
   */
  @bind
  public async make(
    request: Partial<Context> | string,
    setupFn?: (app: Bud) => Promise<unknown>,
  ) {
    if (!this.isRoot) {
      return this.catch(
        InputError.normalize(
          `bud.make: must be called from the root context`,
        ),
      )
    }

    const context: Context = isString(request)
      ? {...this.context, label: request, root: this}
      : {...this.context, ...request, root: this}

    if (isUndefined(context.label)) {
      return this.catch(
        InputError.normalize(`bud.make: context.label must be a string`),
      )
    }

    if (
      !isUndefined(this.context.filter) &&
      !this.context.filter.includes(context.label)
    ) {
      logger
        .scope(`make`)
        .log(
          `skipping child instance based on --filter flag:`,
          context.label,
        )

      return this
    }

    if (!this.children) this.children = {}

    if (this.children[context.label]) {
      return this.catch(
        InputError.normalize(
          `bud.make: child instance ${context.label} already exists`,
        ),
      )
    }

    logger.scope(`make`).log(`instantiating ${context.label}`)

    this.children[context.label] =
      await new this.implementation().initialize({
        ...context,
      })
    if (setupFn) await setupFn(this.children[context.label])

    await this.children[context.label].resolvePromises().catch(this.catch)

    if (typeof request !== `string` && request.dependsOn) {
      this.get(context.label)?.hooks.on(
        `build.dependencies`,
        request.dependsOn,
      )
    }

    return this
  }

  /**
   * Await all promised tasks
   */
  @bind
  public promise(promise: (bud: Bud) => Promise<unknown>): Bud {
    this.promised.push(promise)
    return this
  }

  @bind
  public async resolvePromises() {
    if (this.promised.length === 0) return

    const promises = this.promised

    this.promised = []

    await promises.reduce(async (promised, promise) => {
      await promised
      await promise(this)
    }, Promise.resolve())
  }

  /**
   * Register application services
   */
  @bind
  public async register() {
    await this.executeServiceCallbacks(`register`).catch(this.catch)
  }

  /**
   * Parent {@link Bud} instance
   * @readonly
   */
  public get root(): Bud {
    return this.context?.root ?? this
  }

  /**
   * Set a value on the current instance
   * @param key - key
   * @param value - value
   * @param bind - bind value to current instance (default: true, if bindable)
   */
  @bind
  public set<K extends `${keyof Bud & string}`>(
    key: K,
    input: Bud[K],
    bind: boolean = true,
  ): Bud {
    const toBind = bind && isFunction(input) && `bind` in input

    const value = toBind ? input.bind(this) : input

    Object.defineProperty(this, key, {
      configurable: true,
      enumerable: true,
      value,
      writable: true,
    })

    return this
  }

  /**
   * Log message
   * @deprecated Import logger instance from `@roots/bud-support/logger`
   */
  @bind
  public log(...messages: Array<any>) {
    logger.scope(this.label).log(...messages)
    return this
  }

  /**
   * Log success
   * @deprecated Import logger instance from `@roots/bud-support/logger`
   */
  @bind
  public success(...messages: Array<any>) {
    logger.scope(this.label).log(...messages)
    return this
  }
  /**
   * Log warning
   * @deprecated Import logger instance from `@roots/bud-support/logger`
   */
  @bind
  public warn(...messages: Array<any>) {
    logger.scope(this.label).warn(...messages)
    return this
  }

  /**
   * Log info
   * @deprecated Import logger instance from `@roots/bud-support/logger`
   */
  @bind
  public info(...messages: Array<any>) {
    logger.scope(this.label).info(...messages)
    return this
  }

  /**
   * Log error
   * @deprecated Import logger instance from `@roots/bud-support/logger`
   */
  @bind
  public error(...messages: Array<any>) {
    logger.scope(this.label).error(...messages)
    return this
  }
}
