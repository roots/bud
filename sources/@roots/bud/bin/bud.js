#!/usr/bin/env node
// @ts-check

const {Cli, Builtins} = require('clipanion');
const {BuildCommand, CleanCommand, DevCommand, DoctorCommand, InstallCommand} = require('../lib/cjs/cli/index.js');
const { fs } = require('@roots/bud-support')

const args = process.argv.splice(2);

(async () => {
  const manifest = await fs.readJson(`${__dirname}/../package.json`)
  const name = manifest.name.split('/').pop()

  const cli = new Cli({
    binaryLabel: name,
    binaryName: name,
    binaryVersion:manifest.version,
    enableColors: true,
  })

  cli.register(Builtins.HelpCommand)
  cli.register(Builtins.DefinitionsCommand)
  cli.register(Builtins.VersionCommand)
  cli.register(BuildCommand);
  cli.register(CleanCommand);
  cli.register(DevCommand);
  cli.register(DoctorCommand);
  cli.register(InstallCommand)

  cli.run(args);
})();
