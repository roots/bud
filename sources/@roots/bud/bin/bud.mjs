#!/usr/bin/env node

const run = async () => {
  await import(`../lib/cli/index.js`);
};

run();
