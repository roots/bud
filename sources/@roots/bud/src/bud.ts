import * as Framework from '@roots/bud-framework'

/**
 * ⚡️ Bud
 *
 * @public
 */
export default class Bud extends Framework.Bud {
  public constructor() {
    super(Bud)
    this.implementation = Bud
  }
}
