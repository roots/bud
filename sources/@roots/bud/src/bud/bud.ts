import type Api from '@roots/bud-api'
import type Build from '@roots/bud-build'
import type Cache from '@roots/bud-cache'
import type Compiler from '@roots/bud-compiler'
import type Dashboard from '@roots/bud-dashboard'
import type Extensions from '@roots/bud-extensions'
import {Bud as Framework} from '@roots/bud-framework'
import {Service, ServiceContainer} from '@roots/bud-framework'
import type Hooks from '@roots/bud-hooks'
import type Server from '@roots/bud-server'

/**
 * ## ⚡️ Bud
 *
 * {@link https://bud.js.org/ bud.js}
 */
export class Bud extends Framework {
  public override implementation = Bud

  public declare api: Api

  public declare build: Build

  public declare cache: Cache

  public declare compiler: Compiler

  public declare dashboard: Dashboard

  public declare extensions: Extensions

  public declare hooks: Hooks

  public declare server: Server
}

export {Service, ServiceContainer}
