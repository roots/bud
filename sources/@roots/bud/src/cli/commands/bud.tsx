import type {Context} from '@roots/bud-framework/context'
import type {BaseContext} from '@roots/bud-support/clipanion'
import type {ExecaReturnValue} from '@roots/bud-support/execa'
import type browser from '@roots/bud/cli/flags/browser'

import {join, parse} from 'node:path'
import {env, exit} from 'node:process'

import {Bud} from '@roots/bud-framework'
import chalk from '@roots/bud-support/chalk'
import {Command, Option} from '@roots/bud-support/clipanion'
import {bind} from '@roots/bud-support/decorators/bind'
import {BudError, BudErrorClass} from '@roots/bud-support/errors'
import figures from '@roots/bud-support/figures'
import {Box, render, Static} from '@roots/bud-support/ink'
import isNumber from '@roots/bud-support/lodash/isNumber'
import logger from '@roots/bud-support/logger'
import basedir from '@roots/bud/cli/flags/basedir'
import color from '@roots/bud/cli/flags/color'
import debug from '@roots/bud/cli/flags/debug'
import dry from '@roots/bud/cli/flags/dry'
import filter from '@roots/bud/cli/flags/filter'
import log from '@roots/bud/cli/flags/log'
import mode from '@roots/bud/cli/flags/mode'
import notify from '@roots/bud/cli/flags/notify'
import silent from '@roots/bud/cli/flags/silent'
import storage from '@roots/bud/cli/flags/storage'
import verbose from '@roots/bud/cli/flags/verbose'
import {isset} from '@roots/bud/cli/helpers/isset'
import * as instance from '@roots/bud/instance'

import type {CLIContext} from '../index.js'

import * as Fallback from '../components/Error.js'
import {Menu} from '../components/Menu.js'

export type {BaseContext, Context}
export {Option}

/**
 * {@link Command}
 */
export default class BudCommand extends Command<CLIContext> {
  /**
   * {@link Command.paths}
   */
  public static override paths: Array<Array<string>> = [Command.Default]

  /**
   * {@link Command.usage}
   */
  public static override usage = Command.Usage({
    category: `build`,
    description: `Configurable, extensible build tools for modern single and multi-page web applications`,
    details: `\
      Documentation for this command is available at https://bud.js.org/.

      Any flags which accept a boolean can be negated with the \`--no-\` prefix. For example, \`--no-color\` will disable color output.

      Any command can be exited with \`esc\` or \`ctrl+c\`.

      Common tasks:

        - \`bud build production\` compiles source assets in \`production\` mode.
        - \`bud build development\` compiles source assets in \`development\` mode and updates modules in the browser.
        - \`bud doctor\` checks your system and project for common configuration issues. Try this before making an issue in the bud.js repo.
        - \`bud upgrade\` upgrades bud.js core packages and extensions to the latest version.

      Helpful flags:

        - \`--help\` can be appened to any command for usage information.
        - \`--basedir\` sets the working directory for bud and will be treated as project root.
        - \`--storage\` sets the storage directory. Defaults to the system tmp dir.
        - \`--log\` enables logging. Use \`--log\` in tandem with \`--verbose\` for more detailed output.
        - \`--debug\` enables debug mode. It is very noisy in the terminal but also produces useful output files in the storage directory.
    `,
    examples: [[`Interactive menu of available subcommands`, `$0`]],
  })

  public basedir = basedir

  public declare browser?: typeof browser

  public declare bud?: Bud | undefined

  public color: typeof color = color

  public declare context: CLIContext

  public debug: typeof debug = debug

  public dry = dry(true)

  public filter: typeof filter = filter

  public log = log

  public mode = mode

  public declare notify: typeof notify

  public silent = silent(true)

  public storage = storage

  public verbose: typeof verbose = false

  /**
   * Render static
   */
  public static renderStatic(...children: Array<React.ReactElement>) {
    return render(
      <Static items={children}>
        {(child, id) => <Box key={id}>{child}</Box>}
      </Static>,
    ).unmount()
  }

  /**
   * Execute arbitrary sh command with inherited stdio
   */
  @bind
  public async $(
    bin: string,
    args: Array<string>,
    options = {},
  ): Promise<ExecaReturnValue<string>> {
    const bail = () => setTimeout(exit, 100)

    const {execa} = await import(`@roots/bud-support/execa`)
    const process = execa(bin, args.filter(Boolean), {
      cwd: this.bud.path(),
      encoding: `utf8`,
      env: {NODE_ENV: `development`},
      stdio: `inherit`,
      ...options,
    })
      .on(`data`, data => data && this.context.stdout.write(data))
      .on(
        `error`,
        error => error && this.context.stderr.write(error.message),
      )
      .on(`exit`, bail)
      .on(`disconnect`, bail)
      .on(`close`, bail)

    const result = await process
    return result
  }

  /**
   * Apply context from argv to bud.js instance
   */
  @bind
  public async applyBudArguments(bud: BudCommand[`bud`]) {
    /**
     * Override settings:
     *
     * - when children: all children but not the parent
     * - when no children: the parent;
     */
    const override = async (fn: (bud: Bud) => Promise<any>) => {
      bud.hasChildren
        ? await Promise.all(
            [bud, ...Object.values(bud.children)].map(
              async bud => await fn(bud),
            ),
          )
        : await fn(bud)
    }

    const {context} = bud

    isset(context.input) && bud.setPath(`@src`, context.input)

    isset(context.output) && bud.setPath(`@dist`, context.output)

    isset(context.publicPath) && bud.setPublicPath(context.publicPath)

    isset(context.modules) && bud.setPath(`@modules`, context.modules)

    isset(context.hot) &&
      bud.hooks.on(`dev.middleware.enabled`, (middleware = []) =>
        middleware.filter(key =>
          context.hot === false ? key !== `hot` : context.hot,
        ),
      )

    isset(context.proxy) &&
      bud.hooks.on(
        `dev.middleware.proxy.options.target`,
        new URL(context.proxy),
      )
    isset(context.cache) &&
      (await override(async bud => bud.persist(context.cache)))

    isset(context.minimize) &&
      (await override(async bud => bud.minimize(context.minimize)))

    isset(context.devtool) &&
      (await override(async bud => bud.devtool(context.devtool)))

    isset(context.esm) &&
      (await override(async bud => bud.esm.enable(context.esm)))

    isset(context.html) &&
      (await override(async bud => {
        typeof context.html === `string`
          ? bud.html({template: context.html})
          : bud.html(context.html)
      }))

    isset(context.immutable) &&
      (await override(async bud => bud.cdn.freeze(context.immutable)))

    isset(context.hash) &&
      (await override(async bud => bud.hash(context.hash)))

    isset(context.runtime) &&
      (await override(async bud => bud.runtime(context.runtime)))

    isset(context.splitChunks) &&
      (await override(async bud => bud.splitChunks(context.splitChunks)))

    context.use && (await bud.extensions.add(context.use as any))

    await override(async bud => await bud.promise())
  }

  /**
   * Apply context from env to bud.js instance
   */
  @bind
  public async applyBudEnv(bud: Bud) {
    bud
      .when(bud.env.isString(`APP_MODE`), ({hooks}) =>
        hooks.on(`build.mode`, bud.env.get(`APP_MODE`)),
      )
      .when(
        bud.env.isString(`APP_BASE_PATH`),
        bud => (bud.context.basedir = bud.env.get(`APP_BASE_PATH`)),
      )
      .when(bud.env.isString(`APP_PUBLIC_PATH`), ({hooks}) =>
        hooks.on(
          `build.output.publicPath`,
          bud.env.get(`APP_PUBLIC_PATH`),
        ),
      )
      .when(bud.env.isString(`APP_SRC_PATH`), ({hooks}) =>
        hooks.on(`location.@src`, bud.env.get(`APP_SRC_PATH`)),
      )
      .when(bud.env.isString(`APP_DIST_PATH`), ({hooks}) =>
        hooks.on(`location.@dist`, bud.env.get(`APP_DIST_PATH`)),
      )
      .when(bud.env.isString(`APP_STORAGE_PATH`), ({hooks}) =>
        hooks.on(`location.@storage`, bud.env.get(`APP_STORAGE_PATH`)),
      )
  }

  /**
   * Apply context from manifest to bud.js instance
   */
  @bind
  public async applyBudManifestOptions(bud: Bud) {
    const {bud: manifest} = bud.context.manifest
    if (!manifest) return

    bud
      .when(isset(manifest.publicPath), bud =>
        bud.hooks.on(`build.output.publicPath`, manifest.bud.publicPath),
      )
      .when(isset(manifest.paths?.src), bud =>
        bud.hooks.on(`location.@src`, manifest.bud.paths.src),
      )
      .when(isset(manifest.paths?.dist), bud =>
        bud.hooks.on(`location.@dist`, manifest.bud.paths.dist),
      )
      .when(isset(manifest.paths?.storage), bud =>
        bud.hooks.on(`location.@storage`, manifest.bud.paths.storage),
      )
  }

  /**
   * Binary
   *
   * @remarks
   * String like `node`, `ts-node`, or `bun`. For executing child
   * processes with the same binary as the parent.
   */
  public get bin() {
    return env.BUD_JS_BIN
  }

  /**
   * Handle errors
   */
  @bind
  public override async catch(error: BudErrorClass): Promise<void> {
    if (!error.isBudError) error = BudError.normalize(error)

    if (this.bud?.notifier?.notify) {
      try {
        this.bud.notifier.notify({
          group: this.bud.label,
          message: error?.message,
          subtitle: error?.name ?? `Error`,
          title: this.bud.label ?? `bud.js`,
        })
      } catch (error) {
        logger.warn(error.message ?? error)
      }
    }

    if (this.bud?.dashboard?.instance) {
      this.bud.dashboard.render(error)

      if (this.bud.isProduction) {
        const unmountDashboard = async () =>
          await this.bud.dashboard.instance.waitUntilExit()

        this.bud.compiler?.instance?.close
          ? this.bud.compiler.instance.close(unmountDashboard)
          : await unmountDashboard()
      }
    } else {
      BudCommand.renderStatic(
        <Box flexDirection="column">
          <Fallback.Error error={error} />
        </Box>,
      )
    }

    // fallthrough
    if (!this.bud || this.bud?.isProduction) exit(1)
  }

  /**
   * {@link Command.execute}
   */
  public async execute(): Promise<number | void> {
    render(<Menu cli={this.cli} />)
  }

  /**
   * Make {@link Bud} instance
   */
  @bind
  public async makeBud() {
    const applyCliOptionsCallback = async (bud: Bud) => {
      await Promise.all([
        this.applyBudEnv(bud),
        this.applyBudManifestOptions(bud),
        this.applyBudArguments(bud),
      ]).catch(this.catch)

      await bud.promise().catch(this.catch)
    }

    this.context.dry = this.dry
    this.context.mode = this.mode ?? this.context.mode ?? `production`
    this.context.silent = this.silent

    await import(`../env.${this.context.mode}.js`).catch(this.catch)

    this.bud = instance.get()

    await this.bud.initialize(this.context).catch(this.catch)
    await applyCliOptionsCallback(this.bud).catch(this.catch)

    await this.bud.processConfigs().catch(this.catch)

    await applyCliOptionsCallback(this.bud).catch(this.catch)
    this.bud.hooks.action(`build.before`, applyCliOptionsCallback)

    return this.bud
  }

  /**
   * Run a binary.
   */
  @bind
  public async run(
    path: Array<string>,
    userArgs: Array<string>,
    defaultArgs: Array<string> = [],
  ) {
    let [signifier, ...pathParts] = path

    const binaryPath = await this.bud.module
      .getDirectory(signifier)
      .catch(this.catch)

    if (typeof binaryPath !== `string`) {
      process.exitCode = 3
      throw new Error(`Could not find ${signifier} module`)
    }

    let binary = join(binaryPath, ...pathParts)

    if (!(await this.bud.fs.exists(binary))) {
      const checkedPaths = []
      const parsedParts = parse(join(...pathParts))
      const extensions = [`.js`, `.mjs`, `.cjs`].filter(
        ext => ext !== parsedParts.ext,
      )

      pathParts = await extensions.reduce(async (promise, ext) => {
        const result = await promise
        if (result) return result
        const path = [parsedParts.dir, `${parsedParts.name}${ext}`]
        checkedPaths.push(join(...path))
        if (await this.bud.fs.exists(join(binaryPath, ...path)))
          return path
      }, Promise.resolve(null))

      if (!pathParts) {
        process.exitCode = 2
        throw new Error(
          `Could not find ${signifier} binary\n\nChecked:\n - ${binary}\n - ${checkedPaths
            .map(path => join(binaryPath, path))
            .join(`\n - `)}`,
        )
      }

      binary = join(binaryPath, ...pathParts)
    }

    const binaryArguments = userArgs?.length ? userArgs : defaultArgs

    this.context.stdout.write(
      chalk.dim(
        `${figures.pointerSmall} ${signifier} ${binaryArguments.join(
          ` `,
        )}\n`
          .replace(this.bud.path(`@src`), `@src`)
          .replace(this.bud.path(), ``),
      ),
    )

    const result = await this.$(binary, binaryArguments).catch(() => {})
    const code = result && isNumber(result?.exitCode) ? result.exitCode : 1

    if (code) {
      this.context.stderr.write(
        chalk.red(`${figures.cross} exiting with code ${code}\n`),
      )
      return code
    }

    this.context.stdout.write(chalk.green(`${figures.tick} success\n`))
    return code
  }
}
