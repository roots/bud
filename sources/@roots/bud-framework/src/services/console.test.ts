import {factory} from '@repo/test-kit/bud'
import {beforeEach, describe, expect, it, vi} from 'vitest'
import ConsoleBuffer from './console.js'

describe(`ConsoleBuffer`, () => {
  let bud
  let consoleBuffer: ConsoleBuffer

  beforeEach(async () => {
    bud = await factory({mode: `production`})
    consoleBuffer = new ConsoleBuffer(() => bud)
  })

  it(`should not patch the console in ci`, async () => {
    bud.context.args.ci = true

    await consoleBuffer.register(bud)
    expect(consoleBuffer.restore).not.toBeDefined()
  })

  it(`should store console.log to consoleBuffer.messages`, async () => {
    await consoleBuffer.register(bud)

    console.log(`log-test`)
    console.info(`info-test`)
    console.error(`error-test`)
    console.warn(`warn-test`)

    expect(consoleBuffer.messages).toMatchSnapshot()
  })
})
