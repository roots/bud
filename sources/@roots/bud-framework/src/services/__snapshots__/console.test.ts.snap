// Vitest Snapshot v1

exports[`ConsoleBuffer > should store console.log to consoleBuffer.messages 1`] = `
[
  {
    "message": "log-test",
    "stream": "stdout",
  },
  {
    "message": "info-test",
    "stream": "stdout",
  },
  {
    "message": "error-test",
    "stream": "stderr",
  },
  {
    "message": "warn-test",
    "stream": "stderr",
  },
]
`;
